lib/mantav2-migrate/migrator.js

```diff
@@ -0,0 +1,746 @@
+    let storIds;
+
+    // mkdir -p $delinkDir
+    try {
+        fs.mkdirSync(delinkDir);
+    } catch (mkdirErr) {
+        if (mkdirErr && mkdirErr.code !== 'EEXIST') {
+            cb(mkdirErr);
+            return;
+        }
+    }
+
+    // If already have generated delink scripts, then just validate and return
+    // those.
+    try {
+        filenames = fs.readdirSync(delinkDir);
+    } catch (readdirErr) {
+        cb(readdirErr);
+        return;
+    }
+    scripts.stordelink = filenames
+        .filter(f => /_stordelink.sh$/.test(f))
+        .map(f => path.join(delinkDir, f));
+    scripts.moraydelink = filenames
+        .filter(f => /_moraydelink.sh$/.test(f))
+        .map(f => path.join(delinkDir, f));
+    if (scripts.moraydelink.length > 0) {
+        let missingMorayScripts = [];
+        opts.indexShards.forEach(function(shard) {
+            var expected = path.join(delinkDir, `${shard.host}_moraydelink.sh`);
+            if (scripts.moraydelink.indexOf(expected) === -1) {
+                missingMorayScripts.push(expected);
+            }
+        });
+        if (missingMorayScripts.length > 0) {
+            cb(
+                new VError(
+                    'existing delink scripts are missing the following expected ' +
+                        'moray delink scripts: %s',
+                    missingMorayScripts.join(', ')
+                )
+            );
+        } else {
+            cb(null, scripts, false);
+        }
+        return;
+    }
+
+    // Work through each discovery file and generate the delink info.
+    for (let shardIdx = 0; shardIdx < opts.indexShards.length; shardIdx++) {
+        let shard = opts.indexShards[shardIdx];
+
+        let content;
+        let records;
+        try {
+            content = fs.readFileSync(shard.discoveryFile);
+        } catch (readErr) {
+            cb(readErr);
+            return;
+        }
+        try {
+            records = zlib
+                .gunzipSync(content)
+                .toString('utf8')
+                .trim()
+                .split(/\n/g)
+                .map(line => JSON.parse(line));
+        } catch (parseErr) {
+            cb(new VError(parseErr, 'could not load %s', shard.discoveryFile));
+            return;
+        }
+        log.debug(
+            {discoveryFile: shard.discoveryFile, numRecords: records.length},
+            'loaded discovery file'
+        );
+
+        for (let i = 0; i < records.length; i++) {
+            let obj = records[i];
+            let newUuid = uuidv4();
+
+            if (obj.size === 0) {
+                numZeroByteObjs++;
+            } else {
+                for (let s = 0; s < obj.storageIds.length; s++) {
+                    let storId = obj.storageIds[s];
+                    if (!storCmdsFromStorId.hasOwnProperty(storId)) {
+                        storCmdsFromStorId[storId] = [];
+                    }
+
+                    // Add a `ln <old> <new>` for each storage zone that has
```

Comment:
Just checking, there's no way users can deploy a mako that has a /manta dir split across datasets, right?

Does storCmdsFromStorId somewhere include a set -o errexit to catch potential failures to hardlink in such an (unlikely?) scenario?

Reply:
You can have a dir split across datasets?

The "/manta" dir in a mako is the mountpoint of its delegate dataset, per https://github.com/joyent/manta-mako/blob/f6a0721ec99b42e74288cf7d198ef3cd6f032725/boot/setup.sh#L92-L95



> Does storCmdsFromStorId somewhere include a set -o errexit to catch potential failures to hardlink in such an (unlikely?) scenario?

Yes. The header on the script (from the top of this module) is:

```
const STORDELINK_TEMPLATE = `#!/bin/bash

set -o errexit
set -o pipefail
xtrace_log="/var/tmp/stordelink.$(date -u +%Y%m%dT%H%M%S).xtrace.log"
echo "Writing xtrace output to: \${xtrace_log}" >&2
exec 4>>$xtrace_log
BASH_XTRACEFD=4
set -o xtrace
```